lib/conjur/authenticator.rb

```diff
@@ -0,0 +1,76 @@
+module Conjur
+  # Keeps a fresh Conjur access token in a named file by re-authenticating as needed.
+  class Authenticator
+    TOKEN_LIFESPAN = ( ENV['CONJUR_TOKEN_LIFESPAN'] || 5 * 60 ).to_i.seconds
+    DELAY = ( ENV['CONJUR_TOKEN_REFRESH_DELAY'] || 10 ).to_i.seconds
+
+    attr_reader :authenticate, :filename
+
+    # +authenticate+ should be a proc that authenticates with Conjur and returns an 
+    # access token as a Hash.
+    def initialize authenticate, filename
+      @authenticate = authenticate
+      @filename = filename
+    end
+
+    class << self
+      def default_filename
+        "/etc/conjur-auth/token"
+      end
+
+      # Check the token every +DELAY+ seconds and refresh it if it's out of date. 
+      def run authenticate:, filename: default_filename
+        while true
+          authenticator = Authenticator.new(authenticate, filename)
+          authenticator.refresh unless authenticator.fresh?
+          sleep DELAY
+        end
+      end
+    end
+
+    def fresh?
+      token && (token_age <= TOKEN_LIFESPAN)
+    end
+
+    # Perform atomic replacement of the token
+    def refresh
+      token = authenticate.call
+      require 'fileutils'
+      temp_filename = File.join(directory, "token.#{random}")
+      File.write temp_filename, JSON.pretty_generate(token)
```

Comment:
This should use `Tempfile` class. 

Also we should think about what permissions and ownership the file should have. It should default to the default owner, and `0440` or `0600` (I believe `Tempfile` uses `0600` by default). Or it can use the umask, so default permissions, but then at least warn (or, better, require `--force`) if it's going to end up being world-readable.

Maybe if the target file is already present, it could inherit whatever it had, and warn if there was `o+r` among the permission bits. Of course, that ignores the issue of any security and extended attributes which the op might want to place on the file, which would be gone on replacement, which is another can of worms.

But KISS, YAGNI. I'd say, let's put a comment about that in the code and just make sure to use a reasonably secure default.